[Bug]: [ERROR] [MY-010666] [Server] VillageSQL: 'Failed to load VEF extension 'vsql_complex': failed to load so: error 22

**Describe the bug**
Could be a minor issue or may not be a issue, but using the same datadir for both release and debug builds once I install an extension is not possible. I'm getting this error.

2026-04-06T15:55:03.934011Z 6 [System] [MY-010666] [Server] VillageSQL: 'Validating system schema...'
2026-04-06T15:55:03.940099Z 6 [System] [MY-010666] [Server] VillageSQL: 'System schema verification completed successfully'
2026-04-06T15:55:03.940354Z 6 [ERROR] [MY-010666] [Server] VillageSQL: 'Failed to load VEF extension 'vsql_complex': failed to load so: error 22 (..debug/lib/veb/_expanded/vsql_complex/4b19fe2dd27d19808f093d250fdca9f33de42d755ecae720e234a2f0228f5288/lib/vsql_complex.so: cannot open shared object file: No such file or directory)'
2026-04-06T15:55:03.940363Z 6 [ERROR] [MY-010666] [Server] VillageSQL: 'Extension loading failed'
2026-04-06T15:55:03.940490Z 0 [ERROR] [MY-010666] [Server] VillageSQL: 'Failed to initialize'
2026-04-06T15:55:03.940504Z 0 [ERROR] [MY-010119] [Server] Aborting
2026-04-06T15:55:05.516957Z 0 [System] [MY-010910] [Server] .../debug/runtime_output_directory/mysqld: Shutdown complete (mysqld 8.4.8-villagesql-0.0.4-dev-d72b7c55-debug)  Source distribution.

**To Reproduce**
Steps to reproduce the behavior:
1. Create a fresh datadir with release builds
2. Install the vsql_complex extension using INSTALL EXTENSION vsql_complex;
3. Shutdown the server
4. Restart with debug builds using the same datadir

The error pasted above is seen

Workaround : UNINSTALL the extension and the try, it works

**Expected behavior**
A user should be able to use the datadir both for release and debug builds

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Setup (please complete the following information):**

Distributor ID:	Ubuntu
Description:	Ubuntu 24.04.4 LTS
Release:	24.04
Codename:	noble

commit: d72b7c552382864d4805ba6c9ef9a539e0b9c25a
date: 2026-04-03 09:02:14 +0200
created at: 2026-04-05 11:45:22Z
branch: main
MySQL source 8.4.8-villagesql-0.0.4-dev-d72b7c55
MySQL release maturity LTS

**Additional context**
Add any other context about the problem here.

--- a/mysql-test/suite/villagesql/extension/r/extension_missing_so_restart.result
+++ b/mysql-test/suite/villagesql/extension/r/extension_missing_so_restart.result
@@ -16,11 +16,13 @@ INSERT INTO test_missing_so VALUES (1, '(1.5,2.5)');
 SELECT id, c FROM test_missing_so ORDER BY id;
 id	c
 1	(1.5,2.5)
-Moved .so file away from expansion directory
+Moved .so file away from expansion cache
+Moved .veb file away to prevent re-expansion
 # Restarting server - should fail due to missing .so
 # Server correctly refused to start
 Pattern "Extension loading failed" found
-Restored .so file to expansion directory
+Restored .so file to expansion cache
+Restored .veb file
 # restart: --veb-dir=MYSQLTEST_VARDIR/veb
 # Server started after restoring .so file
 SELECT * FROM INFORMATION_SCHEMA.EXTENSIONS;

--- a/mysql-test/suite/villagesql/extension/r/extension_shared_veb_dir_restart.result
+++ b/mysql-test/suite/villagesql/extension/r/extension_shared_veb_dir_restart.result
@@ -1,0 +1,28 @@
+#
+# Initialize two fresh datadirs
+#
+#
+# Phase 1: Start with data2, install extension
+#
+INSTALL EXTENSION vsql_complex;
+SELECT extension_name, extension_version FROM INFORMATION_SCHEMA.EXTENSIONS;
+EXTENSION_NAME	EXTENSION_VERSION
+vsql_complex	0.0.1
+#
+# Phase 2: Switch to data1 (no extensions installed)
+#
+# data1 has no extensions
+SELECT COUNT(*) AS extension_count FROM INFORMATION_SCHEMA.EXTENSIONS;
+extension_count
+0
+#
+# Phase 3: Switch back to data2 — server should start successfully
+#
+# Extension should still be loaded
+SELECT extension_name, extension_version FROM INFORMATION_SCHEMA.EXTENSIONS;
+EXTENSION_NAME	EXTENSION_VERSION
+vsql_complex	0.0.1
+#
+# Cleanup
+#
+UNINSTALL EXTENSION vsql_complex;

--- a/mysql-test/suite/villagesql/extension/r/extension_veb_basic.result
+++ b/mysql-test/suite/villagesql/extension/r/extension_veb_basic.result
@@ -6,10 +6,10 @@ INSTALL EXTENSION test_veb_ext;
 SELECT * FROM INFORMATION_SCHEMA.EXTENSIONS;
 EXTENSION_NAME	EXTENSION_VERSION
 test_veb_ext	1.2.3
-Verified expansion directory structure: _expanded/test_veb_ext/<SHA256>/
-Manifest exists at: <VEBDIR>/_expanded/test_veb_ext/<SHA256>/manifest.json
-Lib directory exists at: <VEBDIR>/_expanded/test_veb_ext/<SHA256>/lib/
-Shared library exists at: <VEBDIR>/_expanded/test_veb_ext/<SHA256>/lib/test_veb_ext.so
+Verified expansion cache structure: .veb_expansion_cache/test_veb_ext/<SHA256>/
+Manifest exists at: <DATADIR>/.veb_expansion_cache/test_veb_ext/<SHA256>/manifest.json
+Lib directory exists at: <DATADIR>/.veb_expansion_cache/test_veb_ext/<SHA256>/lib/
+Shared library exists at: <DATADIR>/.veb_expansion_cache/test_veb_ext/<SHA256>/lib/test_veb_ext.so
 UNINSTALL EXTENSION test_veb_ext;
 SELECT * FROM INFORMATION_SCHEMA.EXTENSIONS;
 EXTENSION_NAME	EXTENSION_VERSION

--- a/mysql-test/suite/villagesql/extension/t/extension_bad_veb.test
+++ b/mysql-test/suite/villagesql/extension/t/extension_bad_veb.test
@@ -278,8 +278,9 @@ INSTALL EXTENSION symlink_relative;
 SELECT * FROM INFORMATION_SCHEMA.EXTENSIONS;
 
 # Verify no expansion directory was left behind
+let MYSQLD_DATADIR= `SELECT @@datadir`;
 --error 1
---file_exists $veb_dir/_expanded/symlink_relative
+--file_exists $MYSQLD_DATADIR/.veb_expansion_cache/symlink_relative
 
 --echo # Test 2: Symlink with absolute path pointing outside
 --let $extension_name = symlink_absolute
@@ -294,7 +295,7 @@ SELECT * FROM INFORMATION_SCHEMA.EXTENSIONS;
 
 # Verify no expansion directory was left behind
 --error 1
---file_exists $veb_dir/_expanded/symlink_absolute
+--file_exists $MYSQLD_DATADIR/.veb_expansion_cache/symlink_absolute
 
 --echo # Test 3: Deep directory traversal with many ../
 --let $extension_name = deep_traversal

--- a/mysql-test/suite/villagesql/extension/t/extension_missing_so_restart.test
+++ b/mysql-test/suite/villagesql/extension/t/extension_missing_so_restart.test
@@ -1,10 +1,12 @@
 # Test that server refuses to start when an extension's .so file is missing
+# and the .veb file is also unavailable for re-expansion.
 #
 # This test verifies that the server REFUSES to start when an installed
-# extension's .so file is deleted from _expanded. This prevents crashes
+# extension's .so file is deleted from the expansion cache AND the .veb
+# source is also missing (preventing re-expansion). This prevents crashes
 # when accessing tables with custom types.
 #
-# Recovery is tested by restoring the .so file and restarting.
+# Recovery is tested by restoring the files and restarting.
 
 --let $veb_dir = $MYSQLTEST_VARDIR/veb
 --exec mkdir -p $veb_dir
@@ -33,11 +35,12 @@ CREATE TABLE test_missing_so (id INT PRIMARY KEY, c COMPLEX);
 INSERT INTO test_missing_so VALUES (1, '(1.5,2.5)');
 SELECT id, c FROM test_missing_so ORDER BY id;
 
-# Move the .so file away to simulate corruption (we'll restore it later)
+# Move the .so file away AND the .veb file to prevent re-expansion
+let MYSQLD_DATADIR= `SELECT @@datadir`;
 --perl
 use strict;
-my $veb_dir = $ENV{'MYSQLTEST_VARDIR'} . "/veb";
-my $ext_dir = "$veb_dir/_expanded/vsql_complex";
+my $datadir = $ENV{'MYSQLD_DATADIR'};
+my $ext_dir = "$datadir/.veb_expansion_cache/vsql_complex";
 opendir(my $dh, $ext_dir) or die "Can't open $ext_dir: $!";
 my @subdirs = grep { $_ ne '.' && $_ ne '..' && -d "$ext_dir/$_" } readdir($dh);
 closedir($dh);
@@ -48,7 +51,12 @@ closedir($dh);
 my $so_path = "$lib_dir/$so_files[0]";
 my $backup_path = "$lib_dir/$so_files[0].bak";
 rename($so_path, $backup_path) or die "Can't move $so_path to $backup_path: $!";
-print "Moved .so file away from expansion directory\n";
+print "Moved .so file away from expansion cache\n";
+my $veb_dir = $ENV{'MYSQLTEST_VARDIR'} . "/veb";
+my $veb_path = "$veb_dir/vsql_complex.veb";
+my $veb_backup = "$veb_dir/vsql_complex.veb.bak";
+rename($veb_path, $veb_backup) or die "Can't move $veb_path to $veb_backup: $!";
+print "Moved .veb file away to prevent re-expansion\n";
 EOF
 
 --echo # Restarting server - should fail due to missing .so
@@ -75,11 +83,11 @@ let SEARCH_FILE= $MYSQLTEST_VARDIR/log/my_restart.err;
 let SEARCH_PATTERN= Extension loading failed;
 --source include/search_pattern.inc
 
-# Restore the .so file
+# Restore the .so and .veb files
 --perl
 use strict;
-my $veb_dir = $ENV{'MYSQLTEST_VARDIR'} . "/veb";
-my $ext_dir = "$veb_dir/_expanded/vsql_complex";
+my $datadir = $ENV{'MYSQLD_DATADIR'};
+my $ext_dir = "$datadir/.veb_expansion_cache/vsql_complex";
 opendir(my $dh, $ext_dir) or die "Can't open $ext_dir: $!";
 my @subdirs = grep { $_ ne '.' && $_ ne '..' && -d "$ext_dir/$_" } readdir($dh);
 closedir($dh);
@@ -91,7 +99,12 @@ my $backup_path = "$lib_dir/$bak_files[0]";
 my $so_path = $backup_path;
 $so_path =~ s/\.bak$//;
 rename($backup_path, $so_path) or die "Can't restore $backup_path to $so_path: $!";
-print "Restored .so file to expansion directory\n";
+print "Restored .so file to expansion cache\n";
+my $veb_dir = $ENV{'MYSQLTEST_VARDIR'} . "/veb";
+my $veb_backup = "$veb_dir/vsql_complex.veb.bak";
+my $veb_path = "$veb_dir/vsql_complex.veb";
+rename($veb_backup, $veb_path) or die "Can't restore $veb_backup to $veb_path: $!";
+print "Restored .veb file\n";
 EOF
 
 # Restart - server should now load successfully

--- a/mysql-test/suite/villagesql/extension/t/extension_shared_veb_dir_restart.test
+++ b/mysql-test/suite/villagesql/extension/t/extension_shared_veb_dir_restart.test
@@ -1,0 +1,116 @@
+# Test that switching datadirs with a shared veb_dir doesn't break extensions.
+# The expansion cache (.veb_expansion_cache/) lives inside each datadir, so
+# orphan cleanup in one datadir cannot affect another.
+#
+# See: https://github.com/villagesql/villagesql-server/issues/199
+
+# Set up a shared VEB directory
+--let $veb_dir = $MYSQLTEST_VARDIR/veb_shared
+--exec mkdir -p $veb_dir
+
+--let $default_veb_dir = `SELECT @@veb_dir`
+--exec cp $default_veb_dir/vsql_complex.veb $veb_dir/
+
+# Suppress expected error messages
+--disable_query_log
+call mtr.add_suppression("Orphaned expansion directory found but not removed");
+call mtr.add_suppression("Failed to load extension");
+call mtr.add_suppression("Failed to load VEF extension");
+call mtr.add_suppression("Extension loading failed");
+call mtr.add_suppression("Failed to initialize");
+call mtr.add_suppression("dlopen");
+call mtr.add_suppression("no such file");
+call mtr.add_suppression("Aborting");
+call mtr.add_suppression("cannot open shared object");
+--enable_query_log
+
+# Prepare paths
+let BASEDIR= `select @@basedir`;
+let DDIR1=$MYSQL_TMP_DIR/shared_veb_data1;
+let DDIR2=$MYSQL_TMP_DIR/shared_veb_data2;
+let MYSQLD_LOG=$MYSQL_TMP_DIR/shared_veb_restart.log;
+let extra_args=--no-defaults --innodb_dedicated_server=OFF --console --loose-skip-auto_generate_certs --loose-skip-sha256_password_auto_generate_rsa_keys --tls-version= --basedir=$BASEDIR --lc-messages-dir=$MYSQL_SHAREDIR;
+let BOOTSTRAP_SQL=$MYSQL_TMP_DIR/shared_veb_bootstrap.sql;
+
+write_file $BOOTSTRAP_SQL;
+CREATE DATABASE test;
+EOF
+
+--echo #
+--echo # Initialize two fresh datadirs
+--echo #
+
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--shutdown_server
+--source include/wait_until_disconnected.inc
+
+--exec $MYSQLD $extra_args --initialize-insecure --datadir=$DDIR1 --init-file=$BOOTSTRAP_SQL --log-error-verbosity=1 > $MYSQLD_LOG 2>&1
+--exec $MYSQLD $extra_args --initialize-insecure --datadir=$DDIR2 --init-file=$BOOTSTRAP_SQL --log-error-verbosity=1 >> $MYSQLD_LOG 2>&1
+
+--echo #
+--echo # Phase 1: Start with data2, install extension
+--echo #
+
+--replace_result $MYSQL_TMP_DIR MYSQL_TMP_DIR $MYSQLTEST_VARDIR MYSQLTEST_VARDIR
+--exec echo "restart:--datadir=$DDIR2 --veb-dir=$veb_dir" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
+
+INSTALL EXTENSION vsql_complex;
+SELECT extension_name, extension_version FROM INFORMATION_SCHEMA.EXTENSIONS;
+
+--echo #
+--echo # Phase 2: Switch to data1 (no extensions installed)
+--echo #
+
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--shutdown_server
+--source include/wait_until_disconnected.inc
+
+--replace_result $MYSQL_TMP_DIR MYSQL_TMP_DIR $MYSQLTEST_VARDIR MYSQLTEST_VARDIR
+--exec echo "restart:--datadir=$DDIR1 --veb-dir=$veb_dir" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
+
+--echo # data1 has no extensions
+SELECT COUNT(*) AS extension_count FROM INFORMATION_SCHEMA.EXTENSIONS;
+
+--echo #
+--echo # Phase 3: Switch back to data2 — server should start successfully
+--echo #
+
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--shutdown_server
+--source include/wait_until_disconnected.inc
+
+# Capture error output in case of failure
+--error 0,1
+--remove_file $MYSQLTEST_VARDIR/log/shared_veb_restart_err.log
+
+--replace_result $MYSQL_TMP_DIR MYSQL_TMP_DIR $MYSQLTEST_VARDIR MYSQLTEST_VARDIR
+--exec echo "restart:--datadir=$DDIR2 --veb-dir=$veb_dir" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
+
+--echo # Extension should still be loaded
+SELECT extension_name, extension_version FROM INFORMATION_SCHEMA.EXTENSIONS;
+
+--echo #
+--echo # Cleanup
+--echo #
+
+UNINSTALL EXTENSION vsql_complex;
+
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--shutdown_server
+--source include/wait_until_disconnected.inc
+
+--force-rmdir $DDIR1
+--force-rmdir $DDIR2
+--force-rmdir $veb_dir
+--remove_file $BOOTSTRAP_SQL
+--error 0,1
+--remove_file $MYSQLD_LOG
+--error 0,1
+--remove_file $MYSQLTEST_VARDIR/log/shared_veb_restart_err.log
+
+# Restart with original datadir
+--exec echo "restart" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc

--- a/mysql-test/suite/villagesql/extension/t/extension_startup.test
+++ b/mysql-test/suite/villagesql/extension/t/extension_startup.test
@@ -21,11 +21,12 @@ INSTALL EXTENSION startup_test_ext;
 # Verify it's installed
 SELECT * FROM INFORMATION_SCHEMA.EXTENSIONS;
 
-# Create an orphan directory (simulating a leftover from an uninstalled extension)
---exec mkdir -p $veb_dir/_expanded/orphan_ext/0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
+# Create an orphan directory in the expansion cache (simulating a leftover from an uninstalled extension)
+let MYSQLD_DATADIR= `SELECT @@datadir`;
+--exec mkdir -p $MYSQLD_DATADIR/.veb_expansion_cache/orphan_ext/0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
 
 # Verify orphan directory exists before restart
---file_exists $veb_dir/_expanded/orphan_ext/0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
+--file_exists $MYSQLD_DATADIR/.veb_expansion_cache/orphan_ext/0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
 
 --source include/villagesql/binlog_check_end.inc
 

--- a/mysql-test/suite/villagesql/extension/t/extension_veb_basic.test
+++ b/mysql-test/suite/villagesql/extension/t/extension_veb_basic.test
@@ -23,22 +23,23 @@ INSTALL EXTENSION test_veb_ext;
 # Verify version 1.2.3 was read from manifest (not hardcoded as 0.0.1)
 SELECT * FROM INFORMATION_SCHEMA.EXTENSIONS;
 
-# Verify expansion directory structure exists: _expanded/{name}/{sha256}/
---let $veb_dir = `SELECT @@veb_dir`
+# Verify expansion cache directory structure exists:
+#   {datadir}/.veb_expansion_cache/{name}/{sha256}/
+let MYSQLD_DATADIR= `SELECT @@datadir`;
 
-# Check that _expanded directory exists
---file_exists $veb_dir/_expanded
+# Check that .veb_expansion_cache directory exists
+--file_exists $MYSQLD_DATADIR/.veb_expansion_cache
 
 # Check that extension directory exists
---file_exists $veb_dir/_expanded/test_veb_ext
+--file_exists $MYSQLD_DATADIR/.veb_expansion_cache/test_veb_ext
 
 # Find the SHA256 subdirectory (we don't know the exact hash, but there should be exactly one)
 --perl
 use strict;
 use File::Find;
 
-my $veb_dir = $ENV{'MYSQLTEST_VARDIR'} . "/veb";
-my $ext_dir = "$veb_dir/_expanded/test_veb_ext";
+my $datadir = $ENV{'MYSQLD_DATADIR'};
+my $ext_dir = "$datadir/.veb_expansion_cache/test_veb_ext";
 
 # Find all subdirectories (should be exactly one SHA256 hash)
 opendir(my $dh, $ext_dir) or die "Cannot open $ext_dir: $!";
@@ -74,10 +75,10 @@ if (! -f $so_file) {
     die "Expected test_veb_ext.so at: $so_file";
 }
 
-print "Verified expansion directory structure: _expanded/test_veb_ext/<SHA256>/\n";
-print "Manifest exists at: <VEBDIR>/_expanded/test_veb_ext/<SHA256>/manifest.json\n";
-print "Lib directory exists at: <VEBDIR>/_expanded/test_veb_ext/<SHA256>/lib/\n";
-print "Shared library exists at: <VEBDIR>/_expanded/test_veb_ext/<SHA256>/lib/test_veb_ext.so\n";
+print "Verified expansion cache structure: .veb_expansion_cache/test_veb_ext/<SHA256>/\n";
+print "Manifest exists at: <DATADIR>/.veb_expansion_cache/test_veb_ext/<SHA256>/manifest.json\n";
+print "Lib directory exists at: <DATADIR>/.veb_expansion_cache/test_veb_ext/<SHA256>/lib/\n";
+print "Shared library exists at: <DATADIR>/.veb_expansion_cache/test_veb_ext/<SHA256>/lib/test_veb_ext.so\n";
 EOF
 
 # Uninstall the extension (cleanup)

--- a/villagesql/veb/veb_file.cc
+++ b/villagesql/veb/veb_file.cc
@@ -347,10 +347,13 @@ bool expand_veb_to_directory(const std::string &name,
   // Create directory structure: .veb_expansion_cache/,
   // .veb_expansion_cache/{name}/, .veb_expansion_cache/{name}/{sha256}/ Create
   // .veb_expansion_cache/ if needed
+  // TODO(villagesql-windows): On Windows, dotfiles aren't hidden by default.
+  // After creating .veb_expansion_cache/, set the hidden attribute
+  // (e.g., SetFileAttributes with FILE_ATTRIBUTE_HIDDEN).
   if (!my_stat(base_path.c_str(), &dir_stat, MYF(0))) {
     if (my_mkdir(base_path.c_str(), 0755, MYF(0)) != 0) {
-      villagesql_error("Failed to create .veb_expansion_cache directory",
-                       MYF(0));
+      villagesql_error("Failed to create .veb_expansion_cache directory: %s",
+                       MYF(0), base_path.c_str());
       return true;
     }
     LogVSQL(INFORMATION_LEVEL, "Created .veb_expansion_cache directory");

--- a/villagesql/veb/veb_file.cc
+++ b/villagesql/veb/veb_file.cc
@@ -34,6 +34,7 @@
 #include "sql/auth/auth_common.h"
 #include "sql/field.h"
 #include "sql/iterators/row_iterator.h"
+#include "sql/mysqld.h"
 #include "sql/sql_base.h"
 #include "sql/sql_class.h"
 #include "sql/sql_executor.h"
@@ -62,30 +63,37 @@
 namespace villagesql {
 namespace veb {
 
+static std::string get_expansion_cache_base_path() {
+  char path_buf[FN_REFLEN];
+  fn_format(path_buf, ".veb_expansion_cache", mysql_real_data_home, "", 0);
+  return std::string(path_buf);
+}
+
 std::string get_extension_so_path(const std::string &extension_name,
                                   const std::string &sha256) {
-  // Construct path: _expanded/{name}/{sha256}/lib/{name}.so
-  std::string expanded_base = get_veb_path("_expanded");
+  // Construct path:
+  // {datadir}/.veb_expansion_cache/{name}/{sha256}/lib/{name}.so
+  std::string expanded_base = get_expansion_cache_base_path();
   if (expanded_base.empty()) {
     return "";
   }
 
   char path_buf[FN_REFLEN];
 
-  // _expanded/{name}/
+  // .veb_expansion_cache/{name}/
   fn_format(path_buf, extension_name.c_str(), expanded_base.c_str(), "", 0);
   std::string name_dir(path_buf);
 
-  // _expanded/{name}/{sha256}/
+  // .veb_expansion_cache/{name}/{sha256}/
   fn_format(path_buf, sha256.c_str(), name_dir.c_str(), "", 0);
   std::string sha_dir(path_buf);
 
-  // _expanded/{name}/{sha256}/lib/
+  // .veb_expansion_cache/{name}/{sha256}/lib/
   fn_format(path_buf, "lib", sha_dir.c_str(), "", 0);
   std::string lib_dir(path_buf);
 
   // TODO(villagesql-windows): should be .dll on windows.
-  // _expanded/{name}/{sha256}/lib/{name}.so
+  // .veb_expansion_cache/{name}/{sha256}/lib/{name}.so
   std::string so_filename = extension_name + ".so";
   fn_format(path_buf, so_filename.c_str(), lib_dir.c_str(), "", 0);
 
@@ -313,8 +321,8 @@ bool expand_veb_to_directory(const std::string &name,
     return true;
   }
 
-  // Construct expansion path: _expanded/{name}/{sha256}/
-  std::string base_path = get_veb_path("_expanded");
+  // Construct expansion path: {datadir}/.veb_expansion_cache/{name}/{sha256}/
+  std::string base_path = get_expansion_cache_base_path();
 
   char name_dir_buf[FN_REFLEN];
   fn_format(name_dir_buf, name.c_str(), base_path.c_str(), "", 0);
@@ -336,17 +344,19 @@ bool expand_veb_to_directory(const std::string &name,
     return false;  // Already expanded, success
   }
 
-  // Create directory structure: _expanded/, _expanded/{name}/,
-  // _expanded/{name}/{sha256}/ Create _expanded/ if needed
+  // Create directory structure: .veb_expansion_cache/,
+  // .veb_expansion_cache/{name}/, .veb_expansion_cache/{name}/{sha256}/ Create
+  // .veb_expansion_cache/ if needed
   if (!my_stat(base_path.c_str(), &dir_stat, MYF(0))) {
     if (my_mkdir(base_path.c_str(), 0755, MYF(0)) != 0) {
-      villagesql_error("Failed to create _expanded directory", MYF(0));
+      villagesql_error("Failed to create .veb_expansion_cache directory",
+                       MYF(0));
       return true;
     }
-    LogVSQL(INFORMATION_LEVEL, "Created _expanded directory");
+    LogVSQL(INFORMATION_LEVEL, "Created .veb_expansion_cache directory");
   }
 
-  // Create _expanded/{name}/ if needed
+  // Create .veb_expansion_cache/{name}/ if needed
   if (!my_stat(name_dir.c_str(), &dir_stat, MYF(0))) {
     if (my_mkdir(name_dir.c_str(), 0755, MYF(0)) != 0) {
       villagesql_error("Failed to create extension directory for '%s'", MYF(0),
@@ -356,7 +366,7 @@ bool expand_veb_to_directory(const std::string &name,
     LogVSQL(INFORMATION_LEVEL, "Created directory: %s", name_dir.c_str());
   }
 
-  // Create _expanded/{name}/{sha256}/
+  // Create .veb_expansion_cache/{name}/{sha256}/
   if (my_mkdir(expanded_path.c_str(), 0755, MYF(0)) != 0) {
     villagesql_error("Failed to create SHA256 expansion directory", MYF(0));
     return true;
@@ -621,6 +631,30 @@ bool load_installed_extensions(THD *thd) {
         return true;
       }
 
+      // Re-expand VEB if the .so is missing from the expansion cache
+      MY_STAT so_stat;
+      if (!my_stat(so_path.c_str(), &so_stat, MYF(0))) {
+        LogVSQL(INFORMATION_LEVEL,
+                "Extension '%s' .so not found at '%s', re-expanding from VEB",
+                extension_name.c_str(), so_path.c_str());
+        std::string expanded_path;
+        std::string reexpand_sha256;
+        if (expand_veb_to_directory(extension_name, expanded_path,
+                                    reexpand_sha256)) {
+          LogVSQL(ERROR_LEVEL, "Failed to re-expand VEB for extension '%s'",
+                  extension_name.c_str());
+          return true;
+        }
+        if (reexpand_sha256 != sha256) {
+          LogVSQL(ERROR_LEVEL,
+                  "Extension '%s' VEB file has changed: database has SHA256 "
+                  "'%s', current VEB has '%s'",
+                  extension_name.c_str(), sha256.c_str(),
+                  reexpand_sha256.c_str());
+          return true;
+        }
+      }
+
       ExtensionRegistration registration;
       std::string load_error;
       if (load_vef_extension(so_path, extension_name, registration,
@@ -673,20 +707,20 @@ void cleanup_orphaned_expansion_directories(
     const std::set<std::string> &installed_extensions) {
   LogVSQL(INFORMATION_LEVEL, "Cleaning up orphaned expansion directories");
 
-  std::string expanded_base_path = get_veb_path("_expanded");
+  std::string expanded_base_path = get_expansion_cache_base_path();
 
-  // Check if _expanded directory exists
+  // Check if .veb_expansion_cache directory exists
   MY_STAT expanded_stat;
   if (!my_stat(expanded_base_path.c_str(), &expanded_stat, MYF(0)) ||
       !MY_S_ISDIR(expanded_stat.st_mode)) {
-    LogVSQL(INFORMATION_LEVEL, "No _expanded directory found");
+    LogVSQL(INFORMATION_LEVEL, "No .veb_expansion_cache directory found");
     return;
   }
 
-  // Open _expanded directory
+  // Open .veb_expansion_cache directory
   DIR *expanded_dir = opendir(expanded_base_path.c_str());
   if (!expanded_dir) {
-    LogVSQL(WARNING_LEVEL, "Failed to open _expanded directory");
+    LogVSQL(WARNING_LEVEL, "Failed to open .veb_expansion_cache directory");
     return;
   }
 

--- a/villagesql/veb/veb_file.h
+++ b/villagesql/veb/veb_file.h
@@ -29,7 +29,6 @@ namespace veb {
 
 // Get full path to a file/directory within veb_dir
 // e.g., get_veb_path("foo.veb") → "/usr/local/mysql/lib/veb/foo.veb"
-// e.g., get_veb_path("_expanded") → "/usr/local/mysql/lib/veb/_expanded"
 // Returns empty string on error
 std::string get_veb_path(const std::string &filename);
 
@@ -45,17 +44,18 @@ bool calculate_file_sha256(const std::string &filepath, std::string &hash_hex);
 // On success, version is populated with the extension version
 bool load_veb_manifest(const std::string &name, std::string &version);
 
-// Expand VEB archive to directory: {veb_dir}/_expanded/{name}/{sha256}/
+// Expand VEB archive to directory:
+//   {datadir}/.veb_expansion_cache/{name}/{sha256}/
 //
 // Directory structure created:
-//   _expanded/
+//   .veb_expansion_cache/
 //     my_extension/
 //       abc123def.../        (SHA256 of my_extension.veb)
 //         manifest.json
 //         lib/
 //           my_extension.so
 //
-// If _expanded/{name}/{sha256}/ already exists, skips extraction
+// If .veb_expansion_cache/{name}/{sha256}/ already exists, skips extraction
 // Returns false on success, true on error
 // On success, expanded_path contains full path and sha256_hash contains hash
 bool expand_veb_to_directory(const std::string &name,
@@ -67,14 +67,14 @@ bool expand_veb_to_directory(const std::string &name,
 // For each extension:
 //   - Validates the .veb file exists
 //   - Validates manifest.json version matches database
-//   - Does NOT re-expand archive
+//   - Re-expands archive if .so is missing from expansion cache
 //   - Does NOT execute SQL
 // After loading all extensions, calls cleanup_orphaned_expansion_directories()
 // Returns false on success, true on error
 bool load_installed_extensions(THD *thd);
 
 // Remove orphaned expansion directories
-// Scans _expanded/{name}/ subdirectories for each extension
+// Scans .veb_expansion_cache/{name}/ subdirectories for each extension
 // Removes SHA256 subdirectories if the extension is not in installed_extensions
 // set
 // Called during startup loading
@@ -129,8 +129,9 @@ bool load_vef_extension(const std::string &so_path,
 void unload_vef_extension(const ExtensionRegistration &registration);
 
 // Get the path to the .so file for an extension
-// Uses the convention: _expanded/{name}/{sha256}/lib/{name}.so
-// Returns empty string if path cannot be constructed
+// Uses the convention:
+// {datadir}/.veb_expansion_cache/{name}/{sha256}/lib/{name}.so Returns empty
+// string if path cannot be constructed
 std::string get_extension_so_path(const std::string &extension_name,
                                   const std::string &sha256);
 

